### PR TITLE
feat(module:flex): add flex component

### PR DESCRIPTION
**Pull request by ParsaArvanehPA**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [✔] The commit message follows our guidelines: https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/CONTRIBUTING.md#commit
- [✔] Tests for the changes have been added (for bug fixes / features)
- [✔] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[✔] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Application (the showcase website) / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
Added flex component. Already exist in <a href="https://ant.design/components/flex" >Ant design</a>

Issue Number: N/A

## Does this PR introduce a breaking change?
```
[ ] Yes
[✔] No
```

### CODE REVIEW

**Comment by simplejason** [comment]
Thanks for your contribution, I'll add the Chinese translation after the merge, or I can add it directly to your PR  if that's ok with you, we plan to release this feature with v17 :)

**Comment by ParsaArvanehPA** [comment]
> Thanks for your contribution, I'll add the Chinese translation after the merge, or I can add it directly to your PR if that's ok with you, we plan to release this feature with v17 :)


You're welcome. I'm glad to contribute to this project.

The Chinese translation should be fine, since I copied the Mandarin text from the antd site. I hope it matches the meaning and tone of the original text. If you find any errors or inconsistencies, please feel free to directly change them on this PR.

I'm looking forward to seeing this feature in v17 :)

File: components/flex/nz-flex.directive.ts
Comment (HyperLife1119) [inline components/flex/nz-flex.directive.ts:23]: Can we make it a standalone component? Reference: #8037

Comment (ParsaArvanehPA) [reply components/flex/nz-flex.directive.ts:23]: I will check it out and get back to you as soon as possible.

**Comment by OriginRing** [comment]
If you have ideas for long-term maintenance of this component, you can add your information in CODEOWNERS 👍

**Comment by ParsaArvanehPA** [comment]
> If you have ideas for long-term maintenance of this component, you can add your information in CODEOWNERS 👍

I'm glad you asked. I have some ideas for the long-term maintenance of this component. I will add my information in CODEOWNERS soon. Thank you. :)

File: components/flex/nz-flex.directive.ts
Comment (ParsaArvanehPA) [reply components/flex/nz-flex.directive.ts:24]: I have made it a standalone directive. Please check it out and let me know if you have any feedback.

**Comment by HyperLife1119** [comment]
Can we complete the remaining changes within this week? Since we are expecting the v17 release this weekend, if you can make it in time, the flex component will be coming in v17.0.0.

**Comment by ParsaArvanehPA** [comment]
> Can we complete the remaining changes within this week? Since we are expecting the v17 release this weekend, if you can make it in time, the flex component will be coming in v17.0.0.

I am not sure what remaining changes you are referring to. Could you please clarify what changes you have in mind. The current state of this component is identical to antd's flex component.

**Review by HyperLife1119 — CHANGES_REQUESTED** [review]

**Comment by HyperLife1119** [comment]
Oh, sorry. Now you can see what changes need to be made.

**Comment by ParsaArvanehPA** [comment]
> Oh, sorry. Now you can see what changes need to be made.

Yeah, I got them. I will deliver the changes on time.

File: components/flex/nz-flex.directive.ts
Comment (HyperLife1119) [inline components/flex/nz-flex.directive.ts:25]: Let’s fully align ant-design’s flex component parameters.

```ts
@Input() nzVertical: boolean = false;
```

File: components/grid/doc/index.en-US.md
Comment (ParsaArvanehPA) [reply components/grid/doc/index.en-US.md:93]: Yeah np

File: components/flex/style/index.less
Comment (ParsaArvanehPA) [reply components/flex/style/index.less:15]: I initially adopted this approach, but I opted for css variables to simplify the code as most of the styles were similar. I will follow your suggestion and switch the styles by css class.

File: components/flex/style/mixin.less
Comment (HyperLife1119) [inline components/flex/style/mixin.less:1]: The mixins is already imported in index.less so this file is unnecessary

File: components/flex/typings.ts
Comment (ParsaArvanehPA) [reply components/flex/typings.ts:44]: > I don't think we need to enumerate css units because there are too many of them and they are updated all the time.
> 
> `${number}px` | `${number}rem` | `${number}%` `->` `${number}${string}`

I agree with you, as I have not used other css units myself, I assumed they were not useful. However, you are right that we should not restrict the users to these types.

Comment (ParsaArvanehPA) [reply components/flex/typings.ts:44]: > Or we can reuse the types defined by [csstype](https://github.com/frenic/csstype), what do you think?

I appreciate your suggestion, but I am reluctant to use this package as it has not been updated since 2018. However, I find the idea interesting. I have noticed that ng-zorro currently does not have stylelint, and I think supporting linting for styles and strong style typings in the future might be a good idea.

Comment (HyperLife1119) [reply components/flex/typings.ts:44]: Let's change the type to `string`, because I noticed there are more cases here, such as `var(--x)`, `calc(x)`, etc.

File: components/flex/nz-flex.directive.ts
Comment (ParsaArvanehPA) [reply components/flex/nz-flex.directive.ts]: A new input named `nzCenter` has been added, which forces the flex container to align its items both vertically and horizontally.

File: components/flex/typings.ts
Comment (ParsaArvanehPA) [reply components/flex/typings.ts]: I agree with you, I did not consider those cases. Changing the type to string makes sense.

**Comment by ParsaArvanehPA** [comment]
I have completed all the required changes. Please review them and let me know your feedback.
@HyperLife1119

File: components/flex/style/index.less
Comment (HyperLife1119) [inline components/flex/style/index.less:126]: We don't need them because ant-design doesn't have them. We want to be consistent with ant-design as much as possible.

File: components/flex/style/index.less
Comment (HyperLife1119) [inline components/flex/style/index.less:27]: ant-design only supports two directions, horizontal and vertical. We don't need to support so many directions.

Additionally, `nzDirection` needs to be changed to `nzVertical: boolean`

File: components/flex/nz-flex.directive.ts
Comment (HyperLife1119) [reply components/flex/nz-flex.directive.ts]: We don't need nzCenter yet because ant-design doesn't have one yet.

File: components/grid/doc/index.en-US.md
Comment (HyperLife1119) [reply components/grid/doc/index.en-US.md:93]: We need to undo unrelated changes to these two grid documents. If you just want to format them, we should do that in another PR.

**Comment by HyperLife1119** [comment]
It is worth noting that not only the component functions, but also the css class we need to be consistent with ant-design.

File: components/flex/nz-flex.directive.ts
Comment (HyperLife1119) [inline components/flex/nz-flex.directive.ts:48]: We still don't need css variables. All it takes is this:

```
'[style.gap.px]': `nzGap === 'small' ? 8 : nzGap === 'middle' ? 16 : nzGap === 'large' ? 24 : nzGap`,
'[style.flex]': `nzFlex`,
```

File: components/flex/nz-flex.directive.ts
Comment (ParsaArvanehPA) [reply components/flex/nz-flex.directive.ts]: I apologize for the misunderstanding. I thought you wanted to add an input that aligns the items vertically or horizontally. 🤦‍♂️😅

File: components/grid/doc/index.en-US.md
Comment (ParsaArvanehPA) [reply components/grid/doc/index.en-US.md:93]: I apologize for the inconvenience, my prettier kept formatting this file. I will undo the unrelated changes in the next commit.

File: components/flex/style/index.less
Comment (ParsaArvanehPA) [reply components/flex/style/index.less:110]: I appreciate your input. I will fully adhere to the ant-design standards in my future work.

**Comment by ParsaArvanehPA** [comment]
> It is worth noting that not only the component functions, but also the css class we need to be consistent with ant-design.

Thank you for the reminder. I will ensure that both the component functions and the css class are consistent with ant-design in this component and the future components I contribute to.

File: components/flex/nz-flex.directive.ts
Comment (ParsaArvanehPA) [reply components/flex/nz-flex.directive.ts]: I have updated the code to match the antd behavior. It supports both number and string types. In string type, the user can decide the unit of measure. If it is just a number, px is set as the unit of measure by default.

File: components/grid/doc/index.en-US.md
Comment (ParsaArvanehPA) [reply components/grid/doc/index.en-US.md]: done right this time :)

**Comment by ParsaArvanehPA** [comment]
Thank you for your patience. I apologize for the delay. All the mentioned changes are made. Please check them and let me know your feedback.
@HyperLife1119

File: components/flex/nz-flex.directive.ts
Comment (HyperLife1119) [inline components/flex/nz-flex.directive.ts:44]: You cannot use `Number.isInteger()` directly in a template expression. I think we should use a `getter` at this point:

```ts
`[style.gap]`: `gap`

protected get gap() {
  switch (this.nzGap) {
    case 'small': return '8px';
    ...
    default:
      if (typeof this.nzGap === 'number') {
        return this.nzGap + 'px';
      }
      return this.nzGap;
  }
}
```

File: components/flex/style/index.less
Comment (HyperLife1119) [reply components/flex/style/index.less:110]: So let's delete it :)

File: components/flex/flex.module.ts
Comment (HyperLife1119) [inline components/flex/flex.module.ts:12]: I think there should be no need for `CommonModule` here.

**Comment by HyperLife1119** [comment]
We are very close to success :)
@ParsaArvanehPA

File: components/flex/nz-flex.directive.ts
Comment (ParsaArvanehPA) [reply components/flex/nz-flex.directive.ts:44]: 👌

File: components/flex/style/index.less
Comment (HyperLife1119) [inline components/flex/style/index.less:110]: I think you forgot to delete this :)

File: components/flex/nz-flex.directive.ts
Comment (ParsaArvanehPA) [reply components/flex/nz-flex.directive.ts]: This approach has a drawback, as it invokes a method directly in the template expression. This means that the method is called every time a change detection cycle occurs.

see this article(6.1 Template Methods): <a href="https://blog.oasisdigital.com/2017/angular-runtime-performance-guide/">Angular Runtime Performance Guide</a>

Comment (ParsaArvanehPA) [reply components/flex/nz-flex.directive.ts]: I check if an implementation of pipe can prove to be useful.

Comment (HyperLife1119) [reply components/flex/nz-flex.directive.ts]: Don't worry about performance, since this getter requires very little computation and it's fast.

Comment (HyperLife1119) [reply components/flex/nz-flex.directive.ts]: Or you can handle it in ngOnChanges. Because I don't want to add a new pipe here.

File: components/flex/style/index.less
Comment (ParsaArvanehPA) [reply components/flex/style/index.less]: 😬😬😅

File: components/flex/nz-flex.directive.ts
Comment (ParsaArvanehPA) [reply components/flex/nz-flex.directive.ts]: I will use the getter method then

**Comment by ParsaArvanehPA** [comment]
> We are very close to success :)

:DDDD

The last changes have been made :)
check them and let me know if they require any change

**Comment by HyperLife1119** [comment]
Well done, Parsa. Now just need to wait for the CI to come through.

**Comment by ParsaArvanehPA** [comment]
> Well done, Parsa. Now just need to wait for the CI to come through.

Thank you very much, @HyperLifelll9. I am grateful for all the time and energy you spent on this pull request. You have been very helpful and supportive throughout the process. I hope the CI goes well and we can merge this soon :)

**Review by HyperLife1119 — APPROVED** [review]
LGTM!

**Comment by ParsaArvanehPA** [comment]
> LGTM!

@HyperLife1119 
Greetings, I am inquiring about the progress of my pull request that received approval some time ago. I am eager to see my work published and integrated into the main branch. Could you please update me on the expected release date of this feature? I appreciate your prompt response and collaboration.

**Comment by HyperLife1119** [comment]
cc @simplejason

**Comment by c01nd01r** [comment]
Hello, guys!
Is this component already available in the release? I see it in the 'ng-zorro-antd/flex' package, but I don't find it in the documentation on the website.

**Pull request merged**
